### PR TITLE
Implement basic textDocument/documentSymbol

### LINKED ISSUE #87: [feature-request] become a `documentSymbolProvider`
I'm really loving having a language server for Smithy so far. Thanks for making the effort and having one! One missing feature that I've found myself reaching for is `textDocument/documentSymbol`. You can read more about it [here](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol), but especially when I'm working with large smithy files it'd be great to use this to get a nice overview of everything in the file or quickly search for something intelligently in the current document.

### PR DESCRIPTION
**Pull request by kubukoz**
*Issue #, if available:*

Closes #87.

*Description of changes:*

Implements a basic documentSymbolProvider.

Some limitations of this implementation that may be expanded further in future changes:

- flat structure: things like operations could have child symbols, but currently everything is on the top level
- some SymbolKind translations may be made more precise with e.g. member target lookup

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Demo:

image

image

image

### CODE REVIEW

**Comment by srchase** [comment]
Could you fix up the checkstyle issues (star import, etc) so we can confirm the rest of CI is successful?

**Comment by kubukoz** [comment]
done! I can't imagine how you folks can work like this, without an automated rewrite that fixes these issues ;)

File: src/main/java/software/amazon/smithy/lsp/SmithyTextDocumentService.java [hunk 1]
```diff
@@ -388,6 +370,48 @@ private String getLine(List<String> lines, Position position) {
...
+            Model model = project.getModel().unwrap();
+
+            List<DocumentSymbol> symbols = new ArrayList<>();
+
+            URI documentUri = documentIdentifierToUri(params.getTextDocument());
+
+            locations.forEach((shapeId, loc) -> {
+                boolean matchesDocument = URI.create(loc.getUri()).equals(documentUri);
+
+                if (!matchesDocument) return;
```
[the remark this answers is not in the document]
Comment (kubukoz) [reply src/main/java/software/amazon/smithy/lsp/SmithyTextDocumentService.java]: my Java is a bit rusty, but my understanding is that this `return` applies to the lambda in `forEach`, not to `documentSymbol` itself. So ultimately you may still end up with a non-empty list in the output, this just filters the symbols to not include the ones outside your current file.

**Comment by kubukoz** [comment]
Apparently not. Sadly I have no access to a Windows machine so won't be able to play with it. @srchase any hints for resolving this?

**Comment by kubukoz** [comment]
awesome! let me see if it still works as expected on my machine and I'll pull that in.

**Comment by kubukoz** [comment]
looks alright :)

**Review by srchase — APPROVED** [review]

**Pull request merged**